tests/test_tokenization_common.py

```diff
@@ -368,11 +369,21 @@ def test_sentencepiece_tokenize_and_convert_tokens_to_string(self):
         # check if converting back to original text works
         reverse_text = tokenizer.convert_tokens_to_string(tokens)
 
-        if self.test_sentencepiece_ignore_case:
-            reverse_text = reverse_text.lower()
+        # All tokenizers pass this test without the below commented out code.
+        # if self.test_sentencepiece_ignore_case:
+        #     reverse_text = reverse_text.lower()
 
         self.assertEqual(reverse_text, text)
 
+        if isinstance(tokenizer, LayoutXLMTokenizer):
+            return
+
+        input_ids = tokenizer(text).input_ids
+        tokens_including_special = tokenizer.convert_ids_to_tokens(input_ids)
+        reverse_text = tokenizer.convert_tokens_to_string(tokens_including_special)
+
+        self.assertEqual(len(tokenizer.tokenize(reverse_text)), len(input_ids))
+
```

Comment:
What I don't necessarily know how to handle in this `convert_tokens_to_string` function is how to re-add spaces around each special token. I wonder if we should be aligned with what the fast version of the tokenizer would outputs. If it's something that seems relevant to you, I would add this test:
```suggestion

        if self.test_rust_tokenizer:
            rust_tokenizer = self.get_rust_tokenizer()
            special_tokens_string_rust = rust_tokenizer.convert_tokens_to_string(special_tokens)
            self.assertEqual(special_tokens_string, special_tokens_string_rust)
```

Reply:
Initially, I didn't go for the "compare to Rust tokenizer" since not all tokenizers in question have this luxury. Thinking about it now that you mention it, this code will definitely add another layer of reliability to the existing test.

Reply:
Okay, after adding this test, a few previously passing tests failed. This is because of the space you mentioned in another comment. I removed the space, so now the code looks like this:

```python
def convert_tokens_to_string(self, tokens):
    ...
    out_string += self.sp_model.decode(current_sub_tokens) + token  # << a space used to be here
    ...
```

And now all the tests pass, including `T5Tokenizer`.
I will note that I changed calls to `self.sp_model.decode_pieces` to `self.sp_model.decode`, as `decode`, if needed, delegates to `decode_pieces`, and this makes the code uniform, in my opinion.